### PR TITLE
[ADAM-1480] Add switch to disable the fast concat method.

### LINKED ISSUE #1478: Fall back to sequential concat when HDFS concat fails
There's a dizzying array of undocumented cases that will cause HDFS' concat method to fail. The most recent one that I found is running on files in an encrypted HDFS directory:

```
17/04/07 01:45:00 ERROR yarn.ApplicationMaster: User class threw exception: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.HadoopIllegalArgumentException): concat can not be called for files in an encryption zone.
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.concatInternal(FSNamesystem.java:2116)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.concatInt(FSNamesystem.java:2081)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.concat(FSNamesystem.java:2043)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.concat(NameNodeRpcServer.java:814)
	at org.apache.hadoop.hdfs.server.namenode.AuthorizationProviderProxyClientProtocol.concat(AuthorizationProviderProxyClientProtocol.java:280)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.concat(ClientNamenodeProtocolServerSideTranslatorPB.java:562)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:617)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1073)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2086)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2082)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1693)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2080)
```

Yahtzee! Anyways, I'm thinking that the reasonable way to move forward is to fall back on the serial concat code when parallel concatenation fails.

### PR DESCRIPTION
**Pull request by fnothaft**
Resolves #1478.

### CODE REVIEW

File: adam-cli/src/main/scala/org/bdgenomics/adam/cli/ADAM2Vcf.scala [hunk 1]
```diff
@@ -56,6 +56,10 @@ class ADAM2VcfArgs extends Args4jBase with ParquetArgs {
   @Args4jOption(required = false, name = "-single", usage = "Save as a single VCF file.")
   var single: Boolean = false
 
+  @Args4jOption(required = false, name = "-disable_fast_concat",
+    usage = "Disables the parallel file concatenation engine.")
+  var disableFastConcat: Boolean = false
```
Comment (heuermh) [inline adam-cli/src/main/scala/org/bdgenomics/adam/cli/ADAM2Vcf.scala]: Not a fan of double-negatives; is there a better way to do args4j flags that default to true?

File: adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMRDDFunctions.scala [hunk 2]
```diff
@@ -64,6 +64,12 @@ trait ADAMSaveAnyArgs extends SaveArgs {
    * @see asSingleFile
    */
   var deferMerging: Boolean
+
+  /**
+   * If asSingleFile is true and deferMerging is false, disables the use of the
+   * fast file concatenation engine.
+   */
+  var disableFastConcat: Boolean
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMRDDFunctions.scala]: Per this, #1295, and #1438, perhaps we should brainstorm on how to clean up the save methods and ensure consistency.  One or more enums, better SaveArgs classes, a context map, etc.

**Comment by heuermh** [comment]
I'd be willing to merge this without changes.  It adds more pressure towards a future refactoring of save methods for consistency.

**Comment by fnothaft** [comment]
@heuermh with #1487 in the pipe, shall we merge this as is?

**Comment by fnothaft** [comment]
Ping.

**Comment by heuermh** [comment]
> Not a fan of double-negatives; is there a better way to do args4j flags that default to true?

What I'm thinking is something like this
http://stackoverflow.com/questions/15008758/parsing-boolean-values-with-argparse/31347222#31347222

where instead of `-disable` arguments that default to false, we have positive and negative arguments in a mutally exclusive group with the positive one being defaulted to true.  That way the field can be named in the positive and code reads better.  I'm not quite sure how to set up such with Args4j though.

**Comment by fnothaft** [comment]
I don't really see the `-disable` "negative" name as an issue, and I don't have the bandwidth to track down how to do what you're talking about with Args4j. Can we merge this as is?

**Comment by heuermh** [comment]
Args4j can simulate mutually exclusive groups with forbids, however I don't see a way of combining two `--feature` and `--no-feature` style options into a single value as I describe above.

An alternative would be to use `ExplicitBooleanOptionHandler` and default to true, i.e.
```scala
@Args4jOption
  (
    required = false,
    name = "-fast_concat",
    usage = "Enable the parallel file concatenation engine. Defaults to true.",
    handler=classOf[ExplicitBooleanOptionHandler]
  )
var fastConcat: Boolean = true
```

Then these would be true
```
$ adam-submit -- adam2vcf -single foo.adam foo.vcf
$ adam-submit -- adam2vcf -single -fast_concat true foo.adam foo.vcf
```
and this would be false
```
$ adam-submit -- adam2vcf -single -fast_concat false foo.adam foo.vcf
```
and this would fail
```
$ adam-submit -- adam2vcf -single -fast_concat foo.adam foo.vcf
"foo.adam" is not a legal boolean value
 VCF    : The VCF file to convert
 ADAM   : Location to write ADAM Variant data
...
```

Would that be an improvement?

**Comment by heuermh** [comment]
Created #1503 for further discussion

**Comment by heuermh** [comment]
Thank you, @fnothaft!

**Pull request merged**